In [1]:
!pip install spacy

  Obtaining dependency information for spacy from https://files.pythonhosted.org/packages/ca/45/7b43e89b30fe73e32fd8b8ab80c407d326761530a88abd823ec8623772a6/spacy-3.8.4-cp311-cp311-win_amd64.whl.metadata
  Using cached spacy-3.8.4-cp311-cp311-win_amd64.whl.metadata (27 kB)
  Obtaining dependency information for spacy-legacy<3.1.0,>=3.0.11 from https://files.pythonhosted.org/packages/c3/55/12e842c70ff8828e34e543a2c7176dac4da006ca6901c9e8b43efab8bc6b/spacy_legacy-3.0.12-py2.py3-none-any.whl.metadata
  Using cached spacy_legacy-3.0.12-py2.py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for spacy-loggers<2.0.0,>=1.0.0 from https://files.pythonhosted.org/packages/33/78/d1a1a026ef3af911159398c939b1509d5c36fe524c7b644f34a5146c4e16/spacy_loggers-1.0.5-py3-none-any.whl.metadata
  Using cached spacy_loggers-1.0.5-py3-none-any.whl.metadata (23 kB)
  Obtaining dependency information for murmurhash<1.1.0,>=0.28.0 from https://files.pythonhosted.org/packages/de/30/ceb9217cdba72

In [81]:
!python -m spacy download en_core_web_lg

ERROR: Exception:
Traceback (most recent call last):
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\site-packages\pip\_vendor\urllib3\response.py", line 438, in _error_catcher
    yield
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\site-packages\pip\_vendor\urllib3\response.py", line 561, in read
    data = self._fp_read(amt) if not fp_closed else b""
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\site-packages\pip\_vendor\urllib3\response.py", line 527, in _fp_read
    return self._fp.read(amt) if amt is not None else self._fp.read()
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\site-packages\pip\_vendor\cachecontrol\filewrapper.py", line 90, in read
    data = self.__fp.read(amt)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\http\client.py", line 466, in read
    s = self.fp.read(amt)
        ^^^^^^^^^^^^^^^^^
  File "C:\Users\Gayatri_dasari\anaconda3\Lib\socket.py", line 706, in readinto
    ret

     ---------------------------------------- 0.0/400.7 MB ? eta -:--:--
     ---------------------------------------- 0.0/400.7 MB ? eta -:--:--
     -------------------------------------- 0.0/400.7 MB 325.1 kB/s eta 0:20:33
     -------------------------------------- 0.1/400.7 MB 655.4 kB/s eta 0:10:12
     ---------------------------------------- 0.2/400.7 MB 1.2 MB/s eta 0:05:24
     ---------------------------------------- 0.2/400.7 MB 1.4 MB/s eta 0:04:53
     ---------------------------------------- 0.2/400.7 MB 1.4 MB/s eta 0:04:53
     -------------------------------------- 0.3/400.7 MB 896.4 kB/s eta 0:07:27
     -------------------------------------- 0.3/400.7 MB 964.9 kB/s eta 0:06:55
     -------------------------------------- 0.4/400.7 MB 928.4 kB/s eta 0:07:12
     -------------------------------------- 0.4/400.7 MB 897.8 kB/s eta 0:07:26
     -------------------------------------- 0.4/400.7 MB 897.8 kB/s eta 0:07:26
     -------------------------------------- 0.4/400.7 

In [1]:
import spacy
import json
import pandas as pd
from collections import defaultdict
from sklearn.metrics import precision_recall_fscore_support

In [2]:
import os
print("Current working directory:", os.getcwd())
os.chdir('C:/Users/Gayatri_dasari/OneDrive/Desktop/nlp/Assignments/Assignment 1')

Current working directory: C:\Users\Gayatri_dasari\OneDrive\Desktop\nlp\Assignments\Assignment 1


# PHASE 1

## Opening the Manually Annotated JSON File

### loading the spaCy's Large model

In [3]:
nlp=spacy.load('en_core_web_lg')

In [4]:
with open('NER_Tweets_300.json','r') as f:
  Manual_NER = json.load(f)

Manual_NER_list = []

for i in Manual_NER['annotations']:
  if i != None:
    Manual_NER_list.append(
        {
            'sentence': i[0],
            'entities': i[1]['entities']
        }
    )

Manual_NER_list[:10]

[{'sentence': 'The group has been in Colombia since late last year .\r',
  'entities': [[23, 30, 'GPE'], [43, 51, 'DATE']]},
 {'sentence': 'Mr. Ramirez was named defense minister during the two-day long coup which temporarily deposed President Chavez .\r',
  'entities': [[0, 11, 'PERSON'], [50, 57, 'DATE'], [104, 110, 'PERSON']]},
 {'sentence': 'Venezuelan business leader Pedro Carmona , who served as president during the coup , has been granted asylum by Colombia .\r',
  'entities': [[0, 10, 'NORP'], [27, 40, 'PERSON'], [112, 120, 'GPE']]},
 {'sentence': 'Brazilian police have detained two men in connection with the recent theft of nearly $ 70 million from a bank vault in one of the largest bank robberies in history .\r',
  'entities': [[0, 9, 'NORP'], [84, 97, 'MONEY']]},
 {'sentence': 'Authorities say the men were detained near the southeastern city of Belo Horizonte Wednesday , after police stopped their truck loaded with vehicles that were purchased in cash Saturday from a car dea

In [5]:
tweets_df = pd.read_csv('tweets_data.csv')
tweets_df.head()

Sentence
0  The group has been in Colombia since late last...
1  Mr. Ramirez was named defense minister during ...
2  Venezuelan business leader Pedro Carmona , who...
3  Brazilian police have detained two men in conn...
4  Authorities say the men were detained near the...

In [6]:
tweets_df

Sentence
0    The group has been in Colombia since late last...
1    Mr. Ramirez was named defense minister during ...
2    Venezuelan business leader Pedro Carmona , who...
3    Brazilian police have detained two men in conn...
4    Authorities say the men were detained near the...
..                                                 ...
295  The model-turned-activist , currently divorcin...
296  Mills , who lost her left leg below the knee i...
297  She will dance with a sleeve over her prosthes...
298  Heather Mills plans to donate her appearance f...
299  One of the highest-rated shows on U.S. televis...

[300 rows x 1 columns]

In [7]:
tweets_df.describe()

Sentence
count                                                 300
unique                                                300
top     The group has been in Colombia since late last...
freq                                                    1

## Phase2

### Separating the 7 Named Entity Tags only

In [8]:
spacy_output = []
tags = {"PERSON", "NORP", "ORG", "GPE", "LOC", "DATE", "MONEY"}
for sentence in tweets_df['Sentence']:
    doc = nlp(sentence)
    entities = [(ent.text, ent.label_) for ent in doc.ents if ent.label_ in tags]
    spacy_output.append({"sentence": sentence, "entities": entities})

# Display the results
for result in spacy_output:
    print(f"Sentence: {result['sentence']}")
    for entity in result['entities']:
        print(f"  Entity: {entity[0]}, Label: {entity[1]}")
    print("\n")

Sentence: The group has been in Colombia since late last year .
  Entity: Colombia, Label: GPE
  Entity: late last year, Label: DATE


Sentence: Mr. Ramirez was named defense minister during the two-day long coup which temporarily deposed President Chavez .
  Entity: Ramirez, Label: PERSON
  Entity: two-day, Label: DATE
  Entity: Chavez, Label: PERSON


Sentence: Venezuelan business leader Pedro Carmona , who served as president during the coup , has been granted asylum by Colombia .
  Entity: Venezuelan, Label: NORP
  Entity: Pedro Carmona, Label: PERSON
  Entity: Colombia, Label: GPE


Sentence: Brazilian police have detained two men in connection with the recent theft of nearly $ 70 million from a bank vault in one of the largest bank robberies in history .
  Entity: Brazilian, Label: NORP
  Entity: nearly $ 70 million, Label: MONEY


Sentence: Authorities say the men were detained near the southeastern city of Belo Horizonte Wednesday , after police stopped their truck loaded with 

In [9]:
# Named Entities tagged by the SpaCy model
spacy_output

[{'sentence': 'The group has been in Colombia since late last year .',
  'entities': [('Colombia', 'GPE'), ('late last year', 'DATE')]},
 {'sentence': 'Mr. Ramirez was named defense minister during the two-day long coup which temporarily deposed President Chavez .',
  'entities': [('Ramirez', 'PERSON'),
   ('two-day', 'DATE'),
   ('Chavez', 'PERSON')]},
 {'sentence': 'Venezuelan business leader Pedro Carmona , who served as president during the coup , has been granted asylum by Colombia .',
  'entities': [('Venezuelan', 'NORP'),
   ('Pedro Carmona', 'PERSON'),
   ('Colombia', 'GPE')]},
 {'sentence': 'Brazilian police have detained two men in connection with the recent theft of nearly $ 70 million from a bank vault in one of the largest bank robberies in history .',
  'entities': [('Brazilian', 'NORP'), ('nearly $ 70 million', 'MONEY')]},
 {'sentence': 'Authorities say the men were detained near the southeastern city of Belo Horizonte Wednesday , after police stopped their truck loaded 

In [10]:
# Manually tagged Named Entities
Manual_NER_list

[{'sentence': 'The group has been in Colombia since late last year .\r',
  'entities': [[23, 30, 'GPE'], [43, 51, 'DATE']]},
 {'sentence': 'Mr. Ramirez was named defense minister during the two-day long coup which temporarily deposed President Chavez .\r',
  'entities': [[0, 11, 'PERSON'], [50, 57, 'DATE'], [104, 110, 'PERSON']]},
 {'sentence': 'Venezuelan business leader Pedro Carmona , who served as president during the coup , has been granted asylum by Colombia .\r',
  'entities': [[0, 10, 'NORP'], [27, 40, 'PERSON'], [112, 120, 'GPE']]},
 {'sentence': 'Brazilian police have detained two men in connection with the recent theft of nearly $ 70 million from a bank vault in one of the largest bank robberies in history .\r',
  'entities': [[0, 9, 'NORP'], [84, 97, 'MONEY']]},
 {'sentence': 'Authorities say the men were detained near the southeastern city of Belo Horizonte Wednesday , after police stopped their truck loaded with vehicles that were purchased in cash Saturday from a car dea

## Phase 3

### Accuracy Assesment

In [11]:
def retrieve_entity_positions(sentence, entity_text):
    """Find the start and end positions of the entity in the sentence."""
    start = sentence.find(entity_text)
    if start == -1:  # If the entity is not found, then return None
        return None
    end = start + len(entity_text)
    return start, end

## Converting Spacy Structure according to NER

In [12]:
def format_conversion_from_spacy_to_manual(spacy_NER):
    """Convert spaCy format (word, tag) to manual format (start, end, tag)."""
    annotations_after_conversion = []

    for entry in spacy_NER:
        sentence = entry['sentence']
        entities = []

        for entity_text, entity_tag in entry['entities']:
            positions = retrieve_entity_positions(sentence, entity_text) #get each entity start & end index positions just like manual JSON output
            if positions:
                start, end = positions
                entities.append([start, end, entity_tag])

        annotations_after_conversion.append({'sentence': sentence, 'entities': entities})

    return annotations_after_conversion

In [13]:
true_positives = defaultdict(int)
false_positives = defaultdict(int)
false_negatives = defaultdict(int)
false_positives_list = []
false_negatives_list = []
def calculate_metrics(manual_NER, spacy_NER):

    for manual, spacy in zip(manual_NER, spacy_NER):
        # Convert manual annotations into a set of tuples (start, end, label)
        set_manual_entities = set([tuple(entity) for entity in manual['entities']])

        # Convert spaCy annotations into a set of tuples (start, end, label)
        set_spacy_entities = set([tuple(entity) for entity in spacy['entities']])
            
        for entity in set_spacy_entities:
            if entity in set_manual_entities:
                true_positives[entity[2]] += 1
            else:
                false_positives[entity[2]] += 1
                false_positives_list.append(entity)
            
        for entity in set_manual_entities:
            if entity not in set_spacy_entities:
                false_negatives[entity[2]] += 1
                false_negatives_list.append(entity)


In [14]:
# Convert spaCy format to manual format
spacy_after_conversion = format_conversion_from_spacy_to_manual(spacy_output)

# Now calculate TP, FP, FN using the converted spaCy annotations
calculate_metrics(Manual_NER_list, spacy_after_conversion)


In [15]:
print("------------------------------------------------------------------------")
print(f"total_tp: ",true_positives)
print(f"total_fp: ", false_positives)
print(f"total_fn: ",false_negatives)
print("------------------------------------------------------------------------")
print("fp list",false_positives_list)
print("------------------------------------------------------------------------")
print("fn list",false_negatives_list)
print("------------------------------------------------------------------------")

------------------------------------------------------------------------
total_tp:  defaultdict(<class 'int'>, {'DATE': 105, 'PERSON': 56, 'GPE': 154, 'NORP': 88, 'ORG': 23, 'LOC': 4, 'MONEY': 1})
total_fp:  defaultdict(<class 'int'>, {'GPE': 62, 'DATE': 64, 'PERSON': 75, 'MONEY': 7, 'ORG': 63, 'LOC': 13, 'NORP': 29})
total_fn:  defaultdict(<class 'int'>, {'DATE': 62, 'GPE': 79, 'PERSON': 75, 'MONEY': 6, 'ORG': 51, 'NORP': 39, 'LOC': 17})
------------------------------------------------------------------------
fp list [(22, 30, 'GPE'), (37, 51, 'DATE'), (4, 11, 'PERSON'), (78, 97, 'MONEY'), (11, 19, 'DATE'), (89, 98, 'GPE'), (69, 86, 'DATE'), (96, 105, 'DATE'), (45, 51, 'GPE'), (53, 67, 'ORG'), (4, 12, 'PERSON'), (0, 7, 'GPE'), (4, 9, 'PERSON'), (65, 72, 'GPE'), (47, 52, 'PERSON'), (32, 37, 'DATE'), (56, 67, 'ORG'), (75, 80, 'PERSON'), (4, 9, 'PERSON'), (22, 54, 'GPE'), (29, 37, 'DATE'), (95, 108, 'DATE'), (0, 18, 'ORG'), (12, 16, 'ORG'), (69, 73, 'ORG'), (38, 53, 'PERSON'), (100, 116,

In [16]:
labels = set(true_positives.keys()).union(false_positives.keys()).union(false_negatives.keys())

for label in labels:
    tp = true_positives[label]
    fp = false_positives[label]
    fn = false_negatives[label]
    
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
    
    print("------------------------------------------------------------------------")
    print(f"Tag: {label}")
    print(f"  Precision: {precision:.2f}")
    print(f"  Recall:    {recall:.2f}")
    print(f"  F1 Score:  {f1:.2f}\n")

------------------------------------------------------------------------
Tag: ORG
  Precision: 0.27
  Recall:    0.31
  F1 Score:  0.29

------------------------------------------------------------------------
Tag: LOC
  Precision: 0.24
  Recall:    0.19
  F1 Score:  0.21

------------------------------------------------------------------------
Tag: MONEY
  Precision: 0.12
  Recall:    0.14
  F1 Score:  0.13

------------------------------------------------------------------------
Tag: PERSON
  Precision: 0.43
  Recall:    0.43
  F1 Score:  0.43

------------------------------------------------------------------------
Tag: NORP
  Precision: 0.75
  Recall:    0.69
  F1 Score:  0.72

------------------------------------------------------------------------
Tag: GPE
  Precision: 0.71
  Recall:    0.66
  F1 Score:  0.69

------------------------------------------------------------------------
Tag: DATE
  Precision: 0.62
  Recall:    0.63
  F1 Score:  0.62



The performance of the spaCy NER pipeline on different entity categories is varied. Entities like "GPE" (Geopolitical Entities) and "NORP" (Nationalities or Religious/Political Groups) have relatively higher F1 scores of 0.69 and 0.72, respectively, showing that for these entities, precision and recall are better balanced in identifying them. In contrast, categories like "LOC" depict the lowest F1 score at 0.21, while "MONEY" follows at 0.13, showing poor detection and classification in those categories. For "PERSON" and "ORG", it showed a medium performance of 0.43 and 0.29 F1 scores, respectively. These variations underline the model's differing efficacy across entity types, probably because of factors such as the diversity and representation of entities in the training data or due to the inherent complexity of certain categories. Thus, this allows for targeted data augmentation and fine-tuning of the model to improve performance, particularly for the underperforming categories. 